File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    int search(vector<int>& arr, long long target) {
      int left = 0, right = arr.size();
      while(left < right) {
        int middle = left + (right - left) / 2;
        if(arr[middle] < target) {
          left = middle + 1;
        } else {
          right = middle;
        }
      }
      return (int)arr.size() - right;
    }
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        sort(potions.begin(), potions.end());
        vector<int> ans = {};
        int n = potions.size();
        for(int spell: spells) {
          ans.push_back(search(potions, (long long)ceil(success * 1.0 / spell)));
        }
        return ans;
    }
};

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.ts
function lowerBound(potions: number[], success: number): number {
    const n = potions.length;
    let left = 0, right = n;
    
    while (left < right) {
        let mid = Math.floor((left + right) / 2);
        
        if (potions[mid] < success) {
            left = mid + 1;
        }
        else {
            right = mid;
        }
    }
    
    return right;
}

function successfulPairs(spells: number[], potions: number[], success: number): number[] {
    potions.sort((a, b) => a - b);
    const n = potions.length;
    return spells.map(
        (spell) => n - lowerBound(potions, Math.ceil(success / spell))
    );
};